### ISSUE TITLE
p:viewport

### ISSUE THREAD

**Issue by mkraetke** [mag-letex/xprocedit#3]
From what I see, it's currently not possible to establish port connections between steps nested into a compound/multi-container step and its inputs and outputs. In XProc, a compound step is a wrapper for a subpipeline, which include one or more other steps. There are three compound steps: `p:for-each`, `p:group` and `p:viewport`. In contrast, multi-container steps such as `p:try` and `p:choose` specify _alternate_ subpipelines.

Even if you cannot specify input ports for these type of steps explicitely, they take the output from the previous step as input. To specify the start of the subpipeline in the editor, the user would most likely want to connect this implicit input with the first step of the subpipeline.

You can declare zero or more output ports for compound and multi-container steps. If you don't specify an output port explicitely, the output from the last step of the subpipeline is taken as the primary output of the compound/multi-container step. If you specify more than one output port, one port must be marked as primary output, whereas the other output ports are considered as secondary. In this sense, it should be possible to specify multiple output ports for the subpipeline and declare one output as primary.

**Comment by gimsieke** [comment]
`p:try` and `p:choose` are specified in https://github.com/mag-letex/xprocedit/issues/4#issuecomment-514600337

**For `p:viewport`:**

It has no named input and output ports, but for xprocedit, we can say that there is a primary input and a primary output. We can call them `[source]` and `[result]` in xprocedit. The serializing XSLT needs to decide whether to generate `p:viewport-source` for an explicit connection with a non-DRP port, in case the connected ports won’t be serialized adjacently.

(An extra level of trickiness for the serializing XSLT is that it needs to adjust any variable declarations to use also the `p:pipe` that is used as `p:viewport-source`, unless there are explicit connections to other ports.)

In xprocedit’s Meta panel, we don’t need to offer any modifications to the ports. They have immutable properties of `sequence="false"` and `primary="true"`.

**Comment by gimsieke** [comment]
No, it’s also extra tricky in xprocedit. There needs to be an immutable primary input port `current` within the `p:viewport` paper. It represents one piece of the original document. On the outside, the current port is invisible. The primary port of a `p:viewport` step that is part of a subpipeline may be called `[source]`. 

Within the `p:viewport` paper, stubs that give access to other in-scope ports may be added by the user. If a pipeline author wants to refer to the complete document that arrives on the `[source]` port of `p:viewport`, they need to use such a stub that represents a port in the surrounding pipeline.

**Comment by gimsieke** [comment]
`p:for-each` is #13
